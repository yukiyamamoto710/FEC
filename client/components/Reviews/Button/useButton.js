import { useState } from 'react';

export default function useButton(getMoreReviews) {
  const [isAddReview, setIsAddReview] = useState(false);

  const cancelAddReview = () => {
    setIsAddReview(false);
  };

  const handleClickAddReview = () => {
    setIsAddReview(true);
  };

  const handleClickMoreReview = () => {
    getMoreReviews(2);
  };
  return {
    handleClickMoreReview,
    handleClickAddReview,
    cancelAddReview,
    isAddReview,
  };
}
