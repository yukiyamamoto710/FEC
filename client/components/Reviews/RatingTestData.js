const testData1 = {
  characteristics: {
    Comfort: {
      id: 86687,
      value: '2.7500000000000000',
    },
    Quality: {
      id: 86688,
      value: '2.0000000000000000',
    },
    Size: {
      id: 86685,
      value: '2.2500000000000000',
    },
    Width: {
      id: 86686,
      value: '3.2500000000000000',
    },
  },
  product_id: '25821',
  ratings: {
    1: '1',
    4: '1',
    5: '2',
  },
  recommended: {
    false: '1',
    true: '3',
  },
};
const testData2 = {
  characteristics: {
    Size: {
      id: 86685,
      value: '2.2500000000000000',
    },
  },
  product_id: '25721',
  ratings: {
  },
  recommended: {
  },
};

export { testData1, testData2 };
