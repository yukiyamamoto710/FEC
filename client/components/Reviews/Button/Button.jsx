import React from 'react';
import PropTypes from 'prop-types';
import PopOut from '../PopOut/PopOut';
import useButton from './useButton';

const Button = (props) => {
  const {
    rating,
    getMoreReviews,
    addUserReview,
    isMoreReviews,
  } = props;
  const {
    handleClickMoreReview,
    handleClickAddReview,
    cancelAddReview,
    isAddReview,
  } = useButton(getMoreReviews);
  return (
    <div
      data-testid="addmoreBtn"
      className="add_moreBTNContainer"
    >
      {isAddReview
        ? (
          <PopOut
            addUserReview={addUserReview}
            data={rating}
            cancelAddReview={cancelAddReview}
          />
        )
        : null}
      {isMoreReviews === true
        ? (
          <button
            data-testid="moreBtn"
            className="add_moreBTN"
            type="button"
            onClick={handleClickMoreReview}
          >
            MORE REVIEWS
          </button>
        ) : null }
      <button
        data-testid="AddBtn"
        className="add_moreBTN"
        type="button"
        onClick={handleClickAddReview}
      >
        ADD A REVIEWS
      </button>
    </div>
  );
};

Button.propTypes = {
  rating: PropTypes.shape({}),
  isMoreReviews: PropTypes.bool,
  addUserReview: PropTypes.func,
  getMoreReviews: PropTypes.func,
};

Button.defaultProps = {
  rating: {},
  isMoreReviews: true,
  addUserReview: () => (1),
  getMoreReviews: () => (1),
};

export default Button;
