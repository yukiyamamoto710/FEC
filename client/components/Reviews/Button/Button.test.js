/**
* @jest-environment jsdom
*/
import React from 'react';
import renderer from 'react-test-renderer';
import {
  render,
  cleanup,
  fireEvent,
  screen,
} from '@testing-library/react';
import '@testing-library/jest-dom';
import Button from './Button';
import { testData1 } from '../RatingTestData';

afterEach(cleanup);

describe('render correct', () => {
  it('when moreBTNshow be true addBTN should show', () => {
    const test1 = true;
    render(<Button isMoreReviews={test1} />);
    const test = screen.queryByText('MORE REVIEWS');
    expect(test).toBeInTheDocument();
  });

  it('when moreBTNshow be fale addBTN should not show', () => {
    const test2 = false;
    render(<Button isMoreReviews={test2} />);
    const test = screen.queryByText('MORE REVIEWS');
    expect(test).not.toBeInTheDocument();
  });
});

describe('click should trigger func', () => {
  const test1 = true;
  it('when moreBTNshow be true, moreBTN should be able to trigger', () => {
    const testfunc = jest.fn();
    const { getByTestId } = render(
      <Button
        isMoreReviews={test1}
        getMoreReviews={testfunc}
      />,
    );
    fireEvent.click(getByTestId('moreBtn'));
    expect(testfunc).toHaveBeenCalledTimes(1);
    fireEvent.click(getByTestId('moreBtn'));
    expect(testfunc).toHaveBeenCalledTimes(2);
  });

  it('AddBTN should be able to trigger', () => {
    const testfunc = jest.fn();
    const { getByTestId } = render(
      <Button
        isMoreReviews={test1}
        addUserReview={testfunc}
        rating={testData1}
      />,
    );
    fireEvent.click(getByTestId('AddBtn'));
    expect(getByTestId('popout')).toBeInTheDocument();
  });
});

it('when moreBTNshow be true, AddBTN should be able to trigger', () => {
  const testfunc = jest.fn();
  const test = true;
  const { getByTestId } = render(
    <Button
      isMoreReviews={test}
      getMoreReviews={testfunc}
    />,
  );
  fireEvent.click(getByTestId('moreBtn'));
  expect(testfunc).toHaveBeenCalledTimes(1);
});

it('should have default addUserReview and moreReview', () => {
  expect(Button.defaultProps.addUserReview()).toBeDefined();
  expect(Button.defaultProps.getMoreReviews()).toBeDefined();
});

it('matches snapShot', () => {
  const test = true;
  const test1 = jest.fn();
  const test2 = jest.fn();
  const tree = renderer.create(
    <Button
      moreBTNshow={test}
      addReview={test1}
      moreReview={test2}
    />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
