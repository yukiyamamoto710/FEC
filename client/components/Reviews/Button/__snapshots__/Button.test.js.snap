// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapShot 1`] = `
<div
  className="add_moreBTNContainer"
  data-testid="addmoreBtn"
>
  <button
    className="add_moreBTN"
    data-testid="moreBtn"
    onClick={[Function]}
    type="button"
  >
    MORE REVIEWS
  </button>
  <button
    className="add_moreBTN"
    data-testid="AddBtn"
    onClick={[Function]}
    type="button"
  >
    ADD A REVIEWS
  </button>
</div>
`;
